src/basic/extract-word.c

```
@@ -29,54 +29,51 @@
+
+        if (!separators)
+                separators = WHITESPACE;
 
         /* Parses the first word of a string, and returns it in
          * *ret. Removes all quotes in the process. When parsing fails
          * (because of an uneven number of quotes or similar), leaves
          * the pointer *p at the first invalid character. */
 
-        for (;;) {
-                char c = **p;
-
-                if (start) {
-                        if (flags & EXTRACT_DONT_COALESCE_SEPARATORS)
-                                if (!GREEDY_REALLOC(s, allocated, sz+1))
-                                        return -ENOMEM;
+        if (flags & EXTRACT_DONT_COALESCE_SEPARATORS)
+                if (!GREEDY_REALLOC(s, allocated, sz+1))
+                        return -ENOMEM;
 
-                        if (c == 0)
-                                goto finish_force_terminate;
-                        else if (strchr(separators, c)) {
+        for (;; (*p) ++, c = **p) {
+                if (c == 0)
+                        goto finish_force_terminate;
+                else if (strchr(separators, c)) {
+                        if (flags & EXTRACT_DONT_COALESCE_SEPARATORS) {
                                 (*p) ++;
-                                if (flags & EXTRACT_DONT_COALESCE_SEPARATORS)
-                                        goto finish_force_next;
-                                continue;
+                                goto finish_force_next;
                         }
-
+                } else {
                         /* We found a non-blank character, so we will always
                          * want to return a string (even if it is empty),
                          * allocate it here. */
                         if (!GREEDY_REALLOC(s, allocated, sz+1))
                                 return -ENOMEM;
-
-                        start = false;
+                        break;
                 }
+        }
 
+        for (;; (*p) ++, c = **p) {
```

> this is simply a refactor

oh, I see. Sorry.

> in some cases we
might want to ignore errors and resume where we stopped

do you mean the `EXTRACT_*RELAX`-mode? It doesn't return `-EINVAL` on error.